rust/xaynet-core/src/mask/scalar.rs

```diff
@@ -0,0 +1,250 @@
+//! Scalar representation and conversion.
+//!
+//! See the [mask module] documentation since this is a private module anyways.
+//!
+//! [mask module]: ../index.html
+
+use crate::mask::{
+    model::{ratio_to_float, PrimitiveType},
+    PrimitiveCastError,
+};
+use derive_more::{From, Into};
+use num::{
+    clamp,
+    rational::Ratio,
+    traits::{float::FloatCore, ToPrimitive},
+    BigInt,
+    BigUint,
+    One,
+    Unsigned,
+    Zero,
+};
+use serde::{Deserialize, Serialize};
+use std::{convert::TryFrom, fmt::Debug};
+use thiserror::Error;
+
+#[derive(Debug, Clone, PartialEq, Hash, From, Into, Serialize, Deserialize)]
+/// A numerical representation of a machine learning scalar.
+pub struct Scalar(Ratio<BigUint>);
+
+impl Scalar {
+    /// Constructs a new `Scalar` from the given numerator and denominator.
+    pub fn new<U>(numer: U, denom: U) -> Self
+    where
+        U: Unsigned + Into<BigUint>,
+    {
+        Self(Ratio::new(numer.into(), denom.into()))
+    }
+
+    /// Constructs a `Scalar` of unit value.
+    pub fn unit() -> Self {
+        Self(Ratio::one())
+    }
+
+    /// Convenience method for conversion into a non-negative ratio of `BigInt`.
+    pub(crate) fn into_ratio(self) -> Ratio<BigInt> {
+        let (numer, denom) = self.0.into();
+        Ratio::new(numer.into(), denom.into())
+    }
+
+    /// Convenience method for conversion to a non-negative ratio of `BigInt`.
+    pub(crate) fn to_ratio(&self) -> Ratio<BigInt> {
+        let numer = self.0.numer().clone();
+        let denom = self.0.denom().clone();
+        Ratio::new(numer.into(), denom.into())
```

Comment:
with the `From` impl this could be simplified to
```suggestion
        self.clone().into()
```